From b76ad4b47442bbf2b0e4a0bee6a48237a28536f1 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Thu, 23 Apr 2020 13:31:44 +0800
Subject: [PATCH] clk: s32: s32-gen1: correct max size of periphpll_pllodiv
 array

The local periphpll_pllodiv array is only 7 elements, in order to
valid below error due to the array stack-out-of-bounds:

BUG: KASAN: stack-out-of-bounds in s32gen1_clk_plldig+0x1b0/0x230
Read of size 4 at addr ffffff9011627d9c by task swapper/0/0

CPU: 0 PID: 0 Comm: swapper/0 Not tainted 5.2.37-yocto-standard #1
Hardware name: Freescale S32G275 (DT)
Call trace:
 dump_backtrace+0x0/0x1f0
 show_stack+0x24/0x30
 dump_stack+0xbc/0xf4
 print_address_description+0x6c/0x260
 __kasan_report+0x150/0x198
 kasan_report+0xc/0x18
 __asan_load4+0x84/0xa8
 s32gen1_clk_plldig+0x1b0/0x230
 s32gen1_clocks_init+0x5cc/0xbe8
 s32g275_clocks_init+0x20/0x2c
 of_clk_init+0x2e0/0x40c
 time_init+0x1c/0x64
 start_kernel+0x38c/0x534

Memory state around the buggy address:
 ffffff9011627c80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 ffffff9011627d00: f1 f1 f1 f1 f1 f1 04 f2 00 f2 f2 f2 00 f2 f2 f2
>ffffff9011627d80: 00 00 00 04 f3 f3 f3 f3 00 00 00 00 00 00 00 00
                            ^
 ffffff9011627e00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1
 ffffff9011627e80: f1 f1 00 f2 f2 f2 00 00 f3 f3 00 00 00 00 00 00
==================================================================

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/clk/s32/s32-gen1/pll.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/s32/s32-gen1/pll.h b/drivers/clk/s32/s32-gen1/pll.h
index 46342ec1a3c1..a1f4f5ffa4b6 100644
--- a/drivers/clk/s32/s32-gen1/pll.h
+++ b/drivers/clk/s32/s32-gen1/pll.h
@@ -72,7 +72,7 @@
 
 /* Number of PHIs */
 #define ARMPLL_PHI_Nr			(2)
-#define PERIPHPLL_PHI_Nr		(8)
+#define PERIPHPLL_PHI_Nr		(7)
 #define DDRPLL_PHI_Nr			(1)
 #define ACCELPLL_PHI_Nr			(2)
 
@@ -95,7 +95,7 @@
 #define MAX_VCO_RATE			(5000000000)
 #define MIN_VCO_RATE			(1300000000)
 
-#define PHI_MAXNUMBER                   (8)
+#define PHI_MAXNUMBER                   (7)
 #define DFS_MAXNUMBER                   (6)
 
 /* DIV for each PHI. */
-- 
2.17.1

